plugins/modules/files/archive.py

```diff
@@ -214,13 +225,67 @@ def expand_paths(paths):
     return expanded_path, is_globby
 
 
+def matches_exclusion_patterns(path, exclusion_patterns):
+    return any(map(lambda p: fnmatch(path, p), exclusion_patterns))
+
+
+def get_filter(exclusion_patterns, format):
+    def zip_filter(path):
+        return matches_exclusion_patterns(path, exclusion_patterns)
+
+    def tar_filter(tarinfo):
+        return None if matches_exclusion_patterns(tarinfo.name, exclusion_patterns) else tarinfo
+
+    return zip_filter if format == 'zip' or not PY27 else tar_filter
+
+
+def get_archive_contains(format):
+    def archive_contains(archive, name):
+        try:
+            if format == 'zip':
+                archive.getinfo(name)
+            else:
+                archive.getmember(name)
+        except KeyError:
+            return False
+
+        return True
+
+    return archive_contains
+
+
+def get_add_to_archive(format, filter):
+    def add_to_zip_archive(archive_file, path, archive_name):
+        try:
+            if not filter(path):
+                archive_file.write(path, archive_name)
+        except Exception as e:
+            return e
+
+        return ""
```

Comment:
Since an exception object (and not the stringified exception) is returned, this should be `None` and not an empty string:
```suggestion
        return None
```
(Same below.)